src/backend/distributed/planner/intermediate_result_pruning.c

```diff
@@ -0,0 +1,257 @@
+/*-------------------------------------------------------------------------
+ *
+ * intermediate_result_pruning.c
```

Comment:
Let's add the main logic here:

In plain words, each distributed plan pulls the necessary intermediate results to the worker nodes that the plan hits.

For each query that Citus creates a distributed plan, we keep track of the subPlans used in the queryTree, and save it in the distributed plan. Just before Citus executes each subPlan, Citus first keeps track of every worker node that the distributed plan hits, and marks every subPlan should be broadcasted to these nodes. Later, for each subPlan which is a distributed plan, Citus does this operation recursively since these distributed plans may access to different subPlans, and those have to be recorded as well.

Reply:
Done